applications/ShapeOptimizationApplication/python_scripts/analyzer_factory.py

```diff
@@ -11,36 +11,116 @@
+
+# ------------------------------------------------------------------------------
+def _IdentifyInternalResponsesRecursively(responses):
+    internal_responses = []
+
+    for itr in range(responses.size()):
+        response_i = responses[itr]
+
+        if response_i.Has("is_combined"):
+            if response_i["is_combined"].GetBool():
+                internal_responses += _IdentifyInternalResponsesRecursively(response_i["combined_responses"])
+
+        if response_i["use_kratos"].GetBool():
+            identifier = response_i["identifier"].GetString()
+            kratos_settings = response_i["kratos_response_settings"]
+            internal_responses.append([identifier, kratos_settings])
+
+    return internal_responses
+
+# ------------------------------------------------------------------------------
+def _CreateDependencyGraphRecursively(responses):
+    dependency_graph = []
+
+    for itr in range(responses.size()):
+        response_i = responses[itr]
+        identifier = response_i["identifier"].GetString()
+        weight = response_i["weight"].GetDouble()
+
+        if response_i.Has("is_combined"):
+            if response_i["is_combined"].GetBool():
+                sub_dependency_graph = _CreateDependencyGraphRecursively(response_i["combined_responses"])
+                dependency_graph.append((identifier, sub_dependency_graph, weight))
+            else:
+                dependency_graph.append((identifier, [], weight))
+
+    return dependency_graph
 
 # ==============================================================================
 class Analyzer:
     # --------------------------------------------------------------------------
-    def __init__(self, optimization_settings, model_part_controller, external_analyzer):
+    def __init__(self, internal_analyzer, model_part_controller, external_analyzer, dependency_graph = []):
         self.model_part_controller = model_part_controller
+        self.internal_analyzer = internal_analyzer
         self.external_analyzer = external_analyzer
+        self.dependency_graph = dependency_graph
+
+        self.exist_dependencies = False
+        for _, dependencies, _ in dependency_graph:
+            if len(dependencies) > 0:
+                self.exist_dependencies = True
 
-        if self.__IsInternalAnalyzerRequired(optimization_settings):
-            from analyzer_internal import KratosInternalAnalyzer
-            self.internal_analyzer = KratosInternalAnalyzer(optimization_settings, model_part_controller)
-        else:
-            from analyzer_empty import EmptyAnalyzer
-            self.internal_analyzer = EmptyAnalyzer()
-            if isinstance(external_analyzer, EmptyAnalyzer):
-                raise RuntimeError("Neither an internal nor an external analyzer is defined!")
+        self.response_values_filename = "response_values_from_analyzer.csv"
+        self.gradients_norms = {}
+
+        if internal_analyzer.IsEmpty() and external_analyzer.IsEmpty():
```

Comment:
```suggestion
        if isinstance(internal_analyzer, EmptyAnalyzer) and  isinstance(external_analyzer, EmptyAnalyzer):
```

No need for this additional function just for this check.

Reply:
ping

Reply:
Done. I had it like this before, but it did not work for some reason. For sure I had a mistake in the test case. In any case, it runs now.